lib/ladybug/energy_model/model.rb

```diff
@@ -196,15 +204,77 @@ def create_global_construction_set
         end
       end
 
+      def create_schedule_type_limits
+        if @hash[:properties][:energy][:schedule_type_limits]
+          @hash[:properties][:energy][:schedule_type_limits].each do |schedule_type_limit|
+            schedule_type_limit_object = ScheduleTypeLimit.new(schedule_type_limit)
+            schedule_type_limit_object.to_openstudio(@openstudio_model)
+          end
+        end
+      end
+
+      def create_schedules
+        if @hash[:properties][:energy][:schedules]
+          @hash[:properties][:energy][:schedules].each do |schedule|
+            schedule_type = schedule[:type]
+            schedule_object= nil
+
+            case schedule_type
+            when 'ScheduleRulesetAbridged'
+              schedule_object = ScheduleRulesetAbridged.new(schedule)
+            when 'ScheduleFixedIntervalAbridged'
+              schedule_object = ScheduleFixedIntervalAbridged.new(schedule)
+            else
+              raise("Unknown schedule type #{schedule_type}.")
+            end
+            schedule_object.to_openstudio(@openstudio_model)
+          
+          end
+        end
+      end
+
+      def create_space_types
+        if @hash[:properties][:energy][:program_types]
+          @hash[:properties][:energy][:program_types].each do |space_type|
+            space_type_object = SpaceType.new(space_type)
+            space_type_object.to_openstudio(@openstudio_model)
+          end
+        end
+      end 
+
       def create_rooms
-        if @hash[:rooms] 
+        if @hash[:rooms]
+          $room_array_setpoint = [] 
           @hash[:rooms].each do |room|
           room_object = Room.new(room)
-          room_object.to_openstudio(@openstudio_model)
+          openstudio_room = room_object.to_openstudio(@openstudio_model)
+          if room[:properties][:energy][:program_type] && !room[:properties][:energy][:setpoint]
+            $room_array_setpoint << room
+            if $setpoint_array
+              $setpoint_array.each do |setpoint|
```

Comment:
I'm having difficulty understanding what this variable is exactly. Is is an array of all setpoint hashes on ProgramTypes throughout the model?  How does it differ from `room_array_setpoint`?  Why does it appear that you are looping through it inside the for loop that goes through each of the Room hashes? Or am I just not reading the Ruby syntax correctly here?
This is where some in-line comments would be really helpful for understanding what the different sections of your code is doing and what the variables are.

Reply:
@chriswmackey its a global hash created here https://github.com/ladybug-tools-in2/energy-model-measure/blob/7cbaf0701a9a4a45829eef6efdb4186c6bf31ea1/lib/ladybug/energy_model/space_type.rb#L111